lib/remap-lang/src/program.rs

```diff
@@ -110,36 +64,57 @@ impl Program {
             if !constraint.type_def.contains(&program_def)
                 && (!program_def.kind.is_all() || !constraint.allow_any)
             {
-                return Err(RemapError::from(E::from(Error::ResolvesTo(
-                    ResolvesToError(constraint.type_def, program_def),
-                ))));
-            }
+                let want = constraint.type_def.kind;
+                let got = program_def.kind;
+
+                let span = expressions.last().map(|e| e.span()).unwrap_or_default();
 
-            if !constraint.type_def.is_fallible() && type_defs.iter().any(TypeDef::is_fallible) {
-                return Err(RemapError::from(E::from(Error::Fallible)));
+                diagnostics.push(
+                    Diagnostic::error("unexpected return value")
+                        .with_primary(format!("got: {}", got), span)
+                        .with_context(format!("expected: {}", want), span),
+                );
             }
         }
 
-        Ok(Self { expressions })
+        expressions
+            .iter()
+            .filter(|e| e.type_def(&state).is_fallible())
+            .for_each(|e| {
+                diagnostics.push(
+                    Diagnostic::error("unhandled error")
```

Comment:
Sometimes the expression results in a runtime error because the function is a fallible one (eg. `parse_json`). But other times it can because it has been passed a parameter that it can't determine is the correct type.

I think ultimately it would be nice to be able to distinguish between these. So for example with this:

```
 if .ook == true { x = 2 } else { x = "boo" }
 .thing = round(.x)
```

It would be nice for the error reporting to say :

```
error: unhandled error
  ┌─ :2:2
  │
2 │     .thing = round(.x)
  │     ^^^^^^^^^^^^^^^^^^
  │     │
  │     expression can't determine the correct types for the parameters and so may fail if it is passed
  │     an incorrect type.
  │     handle the error case to ensure runtime success
  │
```

However that probably requires a bit of change to the fallibility checks to make fallibilty more than just a simple boolean, which would be out of scope for this PR.

Perhaps would could just change the below error message to say `expression can result in runtime error, possibly because it can't determine the correct types for the parameters`? That would help point the user in the right direction without having to jump to the docs..

Reply:
Actually, as long as we use the `try_boolean()` etc functions in the function implementation, this will already display something like `unexpected type "string" for value`. But I agree that this can use a further refinement, and indeed it requires a bit more work that I considered to be outside the scope of this PR.